### PR TITLE
SDK updates

### LINKED ISSUE #28: SDK Updates
We need to investigate what the IAP support in Defold engine looks like as well as introduce the new session refresh APIs to the client object.

### PR DESCRIPTION
**Pull request by britzl**
Fixes #28

### CODE REVIEW

File: codegen/main.go [hunk 1]
```diff
@@ -361,7 +361,7 @@ local function on_socket_message(socket, message)
 		end
 	elseif message.match_data then
 		if socket.on_matchdata then
-			message.data = b64.decode(message.data)
+			message.match_data.data = b64.decode(message.match_data.data)
```
Comment (britzl) [inline codegen/main.go:364]: A previous commit with this fix was done on the generated code and not on the template...

File: nakama/nakama.lua [hunk 2]
```diff
@@ -2847,17 +2997,23 @@ end
...
 -- Add friends by ID or username to a user's account.
 -- @param client Nakama client
+-- @param ids_arr (table) The account id of a user.
+-- @param usernames_arr (table) The account username of a user.
 -- @param callback Optional callback function. If none is provided the function
 -- is run from within a coroutine and will wait until the call completes and
 -- return the result
 function M.add_friends(
 	client
+	,ids_arr
```
Comment (britzl) [inline nakama/nakama.lua:3007]: How did this work before? It probably didn't....

File: nakama/nakama.lua [hunk 3]
```diff
@@ -235,19 +235,40 @@ function M.create_user_group_list_user_group(
 	}
 end
 
+--------------------------------------------------------------------------------
+--- create_validated_purchase_environment
+-- - UNKNOWN: Unknown environment.  - SANDBOX: Sandbox/test environment.  - PRODUCTION: Production environment.
+function M.create_validated_purchase_environment(
```
Comment (lugehorsam) [inline nakama/nakama.lua:241]: Should this be empty?

File: nakama/nakama.lua [hunk 4]
```diff
@@ -235,19 +235,40 @@ function M.create_user_group_list_user_group(
...
+	)
+	return {
+	}
+end
+
+--------------------------------------------------------------------------------
+--- create_validated_purchase_store
+-- - APPLE_APP_STORE: Apple App Store  - GOOGLE_PLAY_STORE: Google Play Store  - HUAWEI_APP_GALLERY: Huawei App Gallery
+function M.create_validated_purchase_store(
```
Comment (lugehorsam) [inline nakama/nakama.lua:250]: Should this be empty?

File: nakama/nakama.lua [hunk 5]
```diff
@@ -859,6 +898,15 @@ function M.create_api_notification_list(
 	}
 end
 
+--------------------------------------------------------------------------------
+--- create_api_override_operator
+-- Operator that can be used to override the one set in the leaderboard.   - NO_OVERRIDE: Do not override the leaderboard operator.  - BEST: Override the leaderboard operator with BEST.  - SET: Override the leaderboard operator with SET.  - INCREMENT: Override the leaderboard operator with INCREMENT.  - DECREMENT: Override the leaderboard operator with DECREMENT.
+function M.create_api_override_operator(
```
Comment (lugehorsam) [inline nakama/nakama.lua:904]: looks like another empty function

File: nakama/nakama.lua (hunk 3 above)
Comment (britzl) [reply nakama/nakama.lua:241]: No, probably not :-) It's generated code and comes from this spec:

https://github.com/heroiclabs/nakama/blob/master/apigrpc/apigrpc.swagger.json#L3281-L3291

I see that the `enum` property is new and only used here and in the other two places where there's a empty functions. I'll take a look at the code generator and make sure `enum` is handled.

Comment (lugehorsam) [reply nakama/nakama.lua:241]: @britzl yep take a look at this helper: https://github.com/heroiclabs/nakama-dotnet/blob/master/codegen/main.go#L632

Comment (britzl) [reply nakama/nakama.lua]: I'm now detecting enums and handling these appropriately

**Comment by lugehorsam** [comment]
Thanks @britzl 👍  this PR looks good to me.

**Pull request merged**